### PR TITLE
[FIRRTL] add DedupGroupAnnotation

### PR DESCRIPTION
**Pull request by albertchen-sifive**
This implements `DedupGroupAnnotation` as described by @jackkoenig in this comment https://github.com/chipsalliance/firrtl-spec/pull/116#issuecomment-1613923683

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp [hunk 1]
```diff
@@ -1343,18 +1373,26 @@ class DedupPass : public DedupBase<DedupPass> {
...
             return type_isa<RefType>(port.type) && port.isInput();
           }))
         return;
 
       StructuralHasher hasher(hasherConstants);
+      auto dedupGroup =
+          annotations.hasAnnotation(dedupGroupAnnoClass)
+              ? std::optional{annotations.getAnnotation(dedupGroupAnnoClass)
+                                  .getMember<StringAttr>("group")}
+              : std::nullopt;
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/Dedup.cpp:1392]: StringAttr is nullable so can remove std::optinal? 
```suggestion
          annotations.hasAnnotation(dedupGroupAnnoClass)
              ? annotations.getAnnotation(dedupGroupAnnoClass)
                                  .getMember<StringAttr>("group")
              : StringAttr();
```

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp [hunk 2]
```diff
@@ -1343,18 +1373,26 @@ class DedupPass : public DedupBase<DedupPass> {
...
+
       // If the module has input RefType ports, also skip it.
       if (llvm::any_of(module.getPorts(), [&](PortInfo port) {
             return type_isa<RefType>(port.type) && port.isInput();
           }))
         return;
 
       StructuralHasher hasher(hasherConstants);
+      auto dedupGroup =
+          annotations.hasAnnotation(dedupGroupAnnoClass)
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/Dedup.cpp:1389]: Can we remove the annotation?

**Review by uenoku — COMMENTED** [review]
The implementation looks good, can we update `docs/Dialects/FIRRTL/FIRRTLAnnotations.md` as well?

**Comment by albertchen-sifive** [comment]
I added the suggestions. @uenoku could you take another look to make sure I'm not doing anything dumb with null strings or whatever

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp [hunk 3]
```diff
@@ -1470,6 +1511,10 @@ class DedupPass : public DedupBase<DedupPass> {
     if (failed)
       return signalPassFailure();
 
+    for (auto module : modules) {
+      AnnotationSet::removeAnnotations(module, dedupGroupClass);
+    }
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/Dedup.cpp:1516]:
```suggestion
    for (auto module : modules) 
      AnnotationSet::removeAnnotations(module, dedupGroupClass);
    
```

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp [hunk 4]
```diff
@@ -1320,6 +1350,10 @@ class DedupPass : public DedupBase<DedupPass> {
     // Modules annotated with this should not be considered for deduplication.
     auto noDedupClass = StringAttr::get(context, noDedupAnnoClass);
 
+    // Only modules within the same group may be deduplicated.
+    auto dedupGroupClass = StringAttr::get(context, dedupGroupAnnoClass);
+    ;
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/Dedup.cpp:1355]:
```suggestion
```

**Review by uenoku — APPROVED** [review]
LGTM. Do you have thoughts on behaivor when there are multiple DedupGroupAnno? Currently implementation should pick the first DedupGroupAnno so it would not work as expected. I think it wold be fine to just raise an error if there are annotations.

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp [hunk 5]
```diff
@@ -1340,23 +1373,43 @@ class DedupPass : public DedupBase<DedupPass> {
...
 
       StructuralHasher hasher(hasherConstants);
+      llvm::SmallSetVector<StringAttr, 1> groups;
+      for (auto annotation : annotations) {
+        if (annotation.getClass() == dedupGroupClass)
+          groups.insert(annotation.getMember<StringAttr>("group"));
+      }
+      if (groups.size() > 1) {
+        module.emitError("module belongs to multiple dedup groups: ") << groups;
+        failed = true;
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/Dedup.cpp:1401]: I think this is not thread safe. Can we change outer `parallelFor` to `failableParalleFor` and return `failure()` here?

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp [hunk 6]
```diff
@@ -1470,6 +1523,9 @@ class DedupPass : public DedupBase<DedupPass> {
     if (failed)
       return signalPassFailure();
 
+    for (auto module : modules)
+      AnnotationSet::removeAnnotations(module, dedupGroupClass);
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/Dedup.cpp:1527]: It seems `modules` is already invalid here. Probably it should be sufficient to just replace `module` with `circuit.getOps<FModuleOp>()`.

**Review by uenoku — APPROVED** [review]
Thanks!

**Pull request merged**